<h2><font color="#004D7F" size=6>Módulo 2. Análisis de datos</font></h2>



<h1><font color="#004D7F" size=5>2. Estadística descriptiva</font></h1>

<br><br>
<div style="text-align: right">
<font color="#004D7F" size=3>Manuel Castillo-Cara</font><br>
<font color="#004D7F" size=3>Machine Learning con Python</font><br>

---

<h2><font color="#004D7F" size=5>Índice</font></h2>
<a id="indice"></a>

* [1. Introducción](#section1)
    * [1.1. Cargar el conjunto de datos](#section11)
* [2. Funciones de estadística descriptiva](#section2)
    * [2.1. Revisar los datos: _head()_](#section21)
    * [2.2. Dimensiones de los datos: _shape_](#section22)
    * [2.3. Tipo de datos: _dtypes_](#section23)
    * [2.4. Resumen: _describe()_](#section24)
    * [2.5. Distribución entre clases: _groupby('class').size()_](#section25)
    * [2.6. Correlaciones: _corr()_](#section26)
    * [2.7. Asimetría: _skew()_](#section27)

In [ ]:
# Permite ajustar la anchura de la parte útil de la libreta (reduce los márgenes)
from IPython.core.display import display, HTML
display(HTML("<style>.container{ width:98% }</style>"))

---

<a id="section1"></a>
# <font color="#004D7F"> 1. Introducción</font>

Unos de los errores principales que se cometen cuando se comienza a trabajar con machine learning es tomar decisiones directamente a través de los algoritmos sin un análisis previo del conjunto de datos a trabajar. Es importante que entendamos que, un análisis de datos y un buen preprocesamiento de los mismos antes de realizar un tratamiento de modelado, nos llevará no solamente a obtener mejores y más confiables resultados, sino que entenderemos a la perfección nuestro conjunto de datos dando una gran experiencia en el tiempo en la fase de análisis y tratamiento de datos.

<a id="section11"></a>
## <font color="#004D7F">1.1. Cargar el conjunto de datos</font>

Para esta práctica vamos a cargar el conjunto de datos de Pima Indian Diabetes para hacer observaciones con las funciones de estadística descriptiva.

In [1]:
import pandas as pd
filename='data/pima-indians-diabetes.csv'
names=['preg','plas','pres','skin','test','mass','pedi','age','class']
data=pd.read_csv(filename,names=names)
data

preg  plas  pres  skin  test  mass     pedi  age  class
0       6   148    72    35     0  33.6   627.00   50      1
1       1    85    66    29     0  26.6   351.00   31      0
2       8   183    64     0     0  23.3   672.00   32      1
3       1    89    66    23    94  28.1   167.00   21      0
4       0   137    40    35   168  43.1  2288.00   33      1
..    ...   ...   ...   ...   ...   ...      ...  ...    ...
763    10   101    76    48   180  32.9   171.00   63      0
764     2   122    70    27     0  36.8     0.34   27      0
765     5   121    72    23   112  26.2   245.00   30      0
766     1   126    60     0     0  30.1   349.00   47      1
767     1    93    70    31     0  30.4   315.00   23      0

[768 rows x 9 columns]

<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<a id="section2"></a>
# <font color="#004D7F"> 2. Funciones de estadística descriptiva</font>

La inspección de nuestro conjunto de datos es fundamental para poder entender mejor que técnica utilizar; además, nos ayudará a desarrollar nuestra intuición y hacernos preguntas sobre ellos. Las múltiples perspectivas de sus datos lo desafiarán a pensar en los datos de manera diferente.

<a id="section21"></a>
## <font color="#004D7F">2.1. Revisar los datos: _head()_ </font>

Puede revisar las primeras 20 filas de sus datos utilizando la función `head()` en el DataFrame de Pandas. Puede ver que la primera columna enumera el número de fila, lo cual es útil para hacer referencia a una observación específica.

In [2]:
data.head()

preg  plas  pres  skin  test  mass    pedi  age  class
0     6   148    72    35     0  33.6   627.0   50      1
1     1    85    66    29     0  26.6   351.0   31      0
2     8   183    64     0     0  23.3   672.0   32      1
3     1    89    66    23    94  28.1   167.0   21      0
4     0   137    40    35   168  43.1  2288.0   33      1

<a id="section22"></a>
## <font color="#004D7F">2.2. Dimensiones de los datos: _shape_ </font>

Puede revisar la forma y el tamaño de su conjunto de datos imprimiendo la propiedad `shape` en el DataFrame de Pandas. Los resultados se enumeran en filas y luego en columnas. Puede ver que el conjunto de datos tiene 768 filas y 9 columnas

In [4]:
# Dimensions of your data
data.shape #instancias y Caracteristicas

(768, 9)

<a id="section23"></a>
## <font color="#004D7F">2.3. Tipo de datos: _dtypes_ </font>

Puede enumerar los tipos de datos utilizados por el DataFrame para caracterizar cada atributo utilizando la propiedad `dtypes`. Puede ver que la mayoría de los atributos son enteros y que `mass` y `pedi` son tipos de coma flotante.

In [5]:
# Data Types for Each Attribute
data.dtypes

preg       int64
plas       int64
pres       int64
skin       int64
test       int64
mass     float64
pedi     float64
age        int64
class      int64
dtype: object

<a id="section24"></a>
## <font color="#004D7F">2.4. Resumen: _describe()_ </font>

Puedes ver que obtienes muchos datos. Notará algunas llamadas como `pandas.set_option()` utilizada para cambiar la precisión de los números y el ancho preferido de la salida. Esto es para que sea más legible para este ejemplo. Al describir sus datos de esta manera, vale la pena tomarse un tiempo y revisar las observaciones de los resultados.

In [11]:
# Statistical Summary
#pd.set_option('display.width',50)
pd.set_option('precision',3)
data.describe()

preg     plas     pres     skin  \
count  768.000  768.000  768.000  768.000   
mean     3.845  120.895   69.105   20.536   
std      3.370   31.973   19.356   15.952   
min      0.000    0.000    0.000    0.000   
25%      1.000   99.000   62.000    0.000   
50%      3.000  117.000   72.000   23.000   
75%      6.000  140.250   80.000   32.000   
max     17.000  199.000  122.000   99.000   

          test     mass      pedi      age  \
count  768.000  768.000   768.000  768.000   
mean    79.799   31.993   428.235   33.241   
std    115.244    7.884   340.486   11.760   
min      0.000    0.000     0.100   21.000   
25%      0.000   27.300   205.000   24.000   
50%     30.500   32.000   337.000   29.000   
75%    127.250   36.600   591.500   41.000   
max    846.000   67.100  2329.000   81.000   

         class  
count  768.000  
mean     0.349  
std      0.477  
min      0.000  
25%      0.000  
50%      0.000  
75%      1.000  
max      1.000

<a id="section25"></a>
## <font color="#004D7F">2.5. Distribución entre clases: _groupby('class').size()_ </font>

Puede ver que hay casi el doble de observaciones con la clase 0 (sin aparición de diabetes) que con la clase 1 (aparición de diabetes). En este caso podemos observar que las clases se encuentran desbalanceadas por lo que los tenemos que analizar muy los resultados de los algoritmos.

In [12]:
# Class Distribution -> Problemas de clasificacion, binaria o multiclase
data.groupby('class').size()#Saber cuantas instancias tenemos en las clases

class
0    500
1    268
dtype: int64

<a id="section26"></a>
## <font color="#004D7F">2.6. Correlaciones: _corr()_ </font>

Puede usar la función `corr()` para calcular una matriz de correlación. La matriz enumera todos los atributos en la parte superior y lateral, para dar correlación entre todos los pares de atributos (dos veces, porque la matriz es simétrica). Puede ver que la línea diagonal a través de la matriz desde las esquinas superior izquierda a inferior derecha de la matriz muestra una correlación perfecta de cada atributo consigo mismo.

In [14]:
# Pairwise Pearson correlations
data.corr(method='pearson')
#Nos interesa que la correlacion sea lo mas baja posible |corr|~ 0
#Las correlaciones con respecto a la clase debe ser lo mayor posible (+- 0.75) Dependencia lineal

preg   plas   pres   skin   test   mass  \
preg   1.000  0.129  0.141 -0.082 -0.074  0.018   
plas   0.129  1.000  0.153  0.057  0.331  0.221   
pres   0.141  0.153  1.000  0.207  0.089  0.282   
skin  -0.082  0.057  0.207  1.000  0.437  0.393   
test  -0.074  0.331  0.089  0.437  1.000  0.198   
mass   0.018  0.221  0.282  0.393  0.198  1.000   
pedi  -0.026  0.133  0.051  0.154  0.185  0.104   
age    0.544  0.264  0.240 -0.114 -0.042  0.036   
class  0.222  0.467  0.065  0.075  0.131  0.293   

        pedi    age  class  
preg  -0.026  0.544  0.222  
plas   0.133  0.264  0.467  
pres   0.051  0.240  0.065  
skin   0.154 -0.114  0.075  
test   0.185 -0.042  0.131  
mass   0.104  0.036  0.293  
pedi   1.000  0.018  0.177  
age    0.018  1.000  0.238  
class  0.177  0.238  1.000

<a id="section27"></a>
## <font color="#004D7F">2.7. Asimetría: _skew()_ </font>

Puede calcular el sesgo de cada atributo utilizando la función `skew()`. El resultado de inclinación muestra una inclinación positiva (derecha) o negativa (izquierda). Los valores más cercanos a cero muestran menos sesgo.

In [16]:
# Skew for each attribute
data.skew()
data.kurtosis()

preg     0.159
plas     0.641
pres     5.180
skin    -0.520
test     7.214
mass     3.290
pedi     4.029
age      0.643
class   -1.601
dtype: float64

<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<div style="text-align: right"> <font size=6><i class="fa fa-coffee" aria-hidden="true" style="color:#004D7F"></i> </font></div>